emily/handler/src/api/handlers/chainstate.rs

```diff
@@ -25,13 +38,39 @@ use warp::http::StatusCode;
         (status = 500, description = "Internal server error")
     )
 )]
-pub fn get_chainstate(
-    _height: BlockHeight,
+pub async fn get_chainstate_at_height(
+    context: EmilyContext,
+    height: BlockHeight,
 ) -> impl warp::reply::Reply {
-    let response = GetChainstateResponse {
-        ..Default::default()
-    };
-    with_status(json(&response), StatusCode::OK)
+    // Internal handler so `?` can be used correctly while still returning a reply.
+    async fn handler(
+        context: EmilyContext,
+        height: BlockHeight,
+    ) -> Result<impl warp::reply::Reply, Error> {
```

Comment:
The problem with that is that this is taking the parameter from the url, so for this to be changed I'd need to make a struct that grabs this single parameter out of the url by implementing `FromStr`. 

BUT, more importantly, for the requests that take in two different fields we'd need to either make a `struct <Name>RequestField1` and a `struct <Name>RequestField2` or something like that which defeats this model.

Note that I followed basically what you're suggesting in all cases where the input is a request body and not discrete path parameters.

For CreateDeposit:

- route: https://github.com/stacks-network/sbtc/blob/289-feature-dynamodb-data-accessor-structure-for-emily-api/emily/handler/src/api/routes/deposit.rs#L61
- handler: https://github.com/stacks-network/sbtc/blob/289-feature-dynamodb-data-accessor-structure-for-emily-api/emily/handler/src/api/handlers/deposit.rs#L189

Reply:
I think I was suggesting something else, something like this:
```rust
pub async fn get_chainstate_at_height(
    context: EmilyContext,
    height: BlockHeight,
) -> impl warp::reply::Reply {
    let req = GetChainStateAtHeightRequest { context, height };
    super::to_response(req.handler().await)
}

struct GetChainStateAtHeightRequest {
    context: EmilyContext,
    height: BlockHeight,
}

impl GetChainStateAtHeightRequest {
    fn validate(&self) -> Result<(), Error> {
        unimplemented!()
    }
    async fn handler(self) -> Result<impl warp::reply::Reply, Error> {
        self.validate()?;
        // Get chainstate at height - hopefully just one.
        //
        // If there is more than one then there is a state inconsistency. This is potentially
        // okay but the hope then is that the database is actively being repaired.
        let num_to_retrieve_if_multiple = 5;
        ...
    }
}
```
I just wanted to pull the `handler` function out of the function. There might not be any practical reason to organize things this way, at least not yet, but down the line I think there may be.

Reply:
I don't know why the function within a functions stood out to me. I'm okay with not going down the suggested path, what we have here is fine.

Reply:
OH! If that's your line of reasoning I was looking at a different path and just couldn't figure it out. There has to be a way to wrap the function calls within the warp filter such that all errors get routed to replies.

Consider this example:

```rust
fn get_deposit_route(
    context: EmilyContext
) -> impl Filter<Extract = impl warp::Reply, Error = warp::Rejection> + Clone {
    warp::any().map(move || context.clone())
        .and(warp::path!("deposit" / BitcoinTransactionId / BitcoinTransactionOutputIndex ))
        .and(warp::get())
        .then(handle_get_deposit)
}

pub async fn handle_get_deposit(
    context: EmilyContext,
    bitcoin_txid: String,
    bitcoin_tx_output_index: u16,
) -> impl warp::reply::Reply {

    // Add function to handle:
    async fn handler(
        context: EmilyContext,
        bitcoin_txid: BitcoinTransactionId,
        bitcoin_tx_output_index: BitcoinTransactionOutputIndex,
    ) -> Result<impl warp::reply::Reply, Error> {
        // Do something.
    };

    // Then wrap it all up: 
    handler(context, bitcoin_txid, bitcoin_tx_output_index)
        .await
        .map_or_else(Reply::into_response, Reply::into_response)
}
```

There has to be a way to define a function like `with_error_as_reply` similar to my pseudo code.

```rust
async fn with_error_as_reply<T, Args>(
    handle: fn(Args) -> Future< Output = Result<warp::Reply, warp::Reply>, >
) -> fn(Args) -> Future< Output = warp::Reply, >
where
    // Maybe T is the parent filter and we need to determine what the Filter::Extract
    // tuple is (or something like that), and say that Args is the Filter::Extract to the
    // Filter T. Something like this:
    T: Filter,
    // Because we know T is a filter, this is the "Extract" to T's "Filter" (I know this is wrong)
    Args: T::Extract,
{
    |a: Args| async {
        handle(a)
            .await
            .map_or_else(Reply::into_response, Reply::into_response)
    }
}

fn get_deposit_route(
    context: EmilyContext
) -> impl Filter<Extract = impl warp::Reply, Error = warp::Rejection> + Clone {
    warp::any().map(move || context.clone())
        .and(warp::path!("deposit" / BitcoinTransactionId / BitcoinTransactionOutputIndex ))
        .and(warp::get())
        .then(with_error_as_reply(handle_get_deposit))
}

pub async fn handle_get_deposit(
    context: EmilyContext,
    bitcoin_txid: String,
    bitcoin_tx_output_index: u16,
) -> Result<impl warp::reply::Reply, Error> {
    // Do something.
}
```

I'd like to keep it as is if we can't find a way to make this work so each function is self encapsulating, because at some point we'll figure out how to do this and it'll be clearer what we were thinking by having it this uniform way. I also think it allows the handling to be just a single function, which I think it should be, despite the world not wanting it to be so.

I can also suggest a change to warp that makes there a version of this "then" call that routes errors that implement replies to the output.

Reply:
Ahh I see. Yeah I'm okay with what we have here. The second pseudo code example seems possible, but will take more energy than it's worth.